proposals/sealed-interface-freedom.md

```diff
@@ -0,0 +1,170 @@
+
+Add support for `sealed` modality on `interface` type with any visibility. 
+
+> Here and below the word "subclass" includes extending interfaces and implementing objects, unless a more restricted 
+> meaning is implied by the context. 
+
+Other classes or interfaces  _in the same compilation unit_ and _in the package_ may implement or extend the sealed interface, 
+forming a set of direct subclasses that is closed for further extension and is known at compile-time. 
+Direct subclasses may be top-level or nested inside any number of other named classes, named interfaces, or named objects in the same package.
+Subclass of a sealed interface can be `enum class`, but `annotation class` subclass is not allowed. 
+
+> Kotlin `annotation class` is not allowed to implement any interfaces whatsoever.  
+ 
+Subclasses are allowed to have any visibility as long as they are otherwise compatible with normal Kotlin inheritance rules. 
+It is an error to subclass a sealed interface in a different compilation unit or in a different package.
+Subclasses of a sealed interface must have a proper qualified name &mdash; they _cannot be_ local nor anonymous objects.
+`fun interface` is not allowed to be `sealed`, because the primary way of implementing a `fun interface` is via
+functional conversion that creates anonymous instances, too.   
+
+> The restrictions above allow more freedom than a `sealed class` currently provides, owning to the use-cases that
+> has surfaced during practical use of sealed classes. However, these restrictions are compatible with restrictions 
+> on sealed classes that are introduced by JVM specification.
+>
+> Unlike Java, Kotlin does not require any kind of `permits` annotation even when subclasses are specified in another 
+> file, which honors Kotlin tradition of avoiding source-code repetition of information that could be inferred by 
+> the compiler. Just like return types of functions are optionally specified, we do reserve a possibility that, 
+> in the future, we might add an optional way to specify some kind of `permits` list, but we are reluctant to do this 
+> from day one due to the lack of evidence that this will significantly help to substantiate additional syntactic 
+> feature in the language.  We plan to address the needs of seeing the list of subclasses by other tools, such as IDE 
+> and an API compatibility validator.
+
+An `interface` in Kotlin is `open` by default. Among the class _modalities_ supported by Kotlin 
+(`open`, `sealed`, `abstract`, and `final`) only `open` have been allowed to be (redundantly) specified for an interface 
+and this proposal adds support for a `sealed` modality. Specifying both `sealed` and `open` modalities together is an error.
+
+> Unlike Java, Kotlin subclasses of a sealed interface don't have special rules for their modality. An implementing 
+> class is `final` by default and extending interface is `open` by default, just like otherwise normal Kotlin class
```

Comment:
This should also probably be abstract
```suggestion
> class is `final` by default and extending interface is `abstract` by default, just like otherwise normal Kotlin class
```